File: packages/jarvis/src/commands/fs.ts
import { invoke } from "@tauri-apps/api/core"
import { generateJarvisPluginCommand } from "./common"

export function pathExists(path: string): Promise<boolean> {
	return invoke(generateJarvisPluginCommand("path_exists"), { path })
}

/**
 * This command is built into Jarvis App
 * Used to decompress a tarball file
 * @param path
 * @param destinationFolder
 * @param options
 * @returns
 */
export function decompressTarball(
	path: string,
	destinationFolder: string,
	options?: {
		overwrite?: boolean
	}
): Promise<string> {
	return invoke(generateJarvisPluginCommand("decompress_tarball"), {
		path,
		destinationFolder,
		overwrite: options?.overwrite ?? false
	})
}

/**
 * Compress a given directory into a tarball file
 * @param srcDir Directory to compress
 * @param destFile destination file, should end with .tar.gz or .tgz
 * @param options
 * @returns
 */
export function compressTarball(
	srcDir: string,
	destFile: string,
	options?: {
		overwrite?: boolean
	}
): Promise<string> {
	return invoke(generateJarvisPluginCommand("compress_tarball"), {
		srcDir,
		destFile,
		overwrite: options?.overwrite ?? false
	})
}

/**
 *
 * @param path Path of file to unzip
 * @param destinationFolder where to unzip the file
 * @param options use overwrite to overwrite existing files
 * @returns
 */
export function unzip(
	path: string,
	destinationFolder: string,
	options?: {
		overwrite?: boolean
	}
): Promise<void> {
	return invoke(generateJarvisPluginCommand("unzip"), {
		path,
		destinationFolder,
		overwrite: options?.overwrite ?? false
	})
}

File: packages/tauri-plugins/jarvis/src/commands/fs.rs
use crate::utils;
use std::path::PathBuf;

#[tauri::command]
pub async fn path_exists(path: String) -> Result<bool, String> {
    let exists = PathBuf::from(path.clone()).exists();
    Ok(exists)
}

#[tauri::command]
pub async fn decompress_tarball(
    path: PathBuf,
    destination_folder: PathBuf,
    overwrite: bool,
) -> Result<PathBuf, String> {
    utils::fs::decompress_tarball(&path, &destination_folder, overwrite).map_err(|e| e.to_string())
}

/// Compress a directory into a tarball
/// Both `src_dir` and `dest_file` must be be absolute paths
#[tauri::command]
pub async fn compress_tarball(
    src_dir: PathBuf,
    dest_file: PathBuf,
    overwrite: bool,
) -> Result<PathBuf, String> {
    utils::fs::compress_tarball(&src_dir, &dest_file, overwrite).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn unzip(
    path: PathBuf,
    destination_folder: PathBuf,
    overwrite: bool,
) -> Result<(), String> {
    utils::fs::unzip(&path, &destination_folder, overwrite).map_err(|err| err.to_string())
}
